docs/enhancements/install-strategy.md

```diff
@@ -0,0 +1,196 @@
+Retrying | Unknown initially, True if we've hit an error and are retrying, False if we've given up or don't support retries. Message should indicate attempt x of y.
+
+These interface conditions will be copied back onto ClusterDeployment by Hive controllers with a "Provision" prefix, the install strategy controller does not need to worry about this and in general should never write to ClusterDeployment. example: Failed becomes ProvisionFailed (which exists today on ClusterDeployment for this purpose)
+
+Install strategy controllers can add their own conditions, but these will not transfer back to the ClusterDeployment. UIs around this process may need to show install strategy conditions explicitly.
+
+The core install strategy CRDs and controllers discussed in this document should live in Hive. However, in theory, it would be possible for an external application to implement it's own install strategy outside of Hive and still interface with ClusterDeployment using this system.
+
+### Migration Plan
+
+No breakage for current installation API is planned. (with exception of as yet unreleased Agent install strategy)
+
+Initial goal would be to breakout ClusterDeployment.Spec.InstallStrategy.Agent. The only implementation here is agent based and this is hidden behind an alpha feature gate, and not yet released to customers. Replace this with a new hive.openshift.io.AgentInstallStrategy CRD, and a ClusterDeployment.Spec.InstallStrategyRef link. Refactor the Agent controllers (currently living in the assisted service repo, coming to hive in future) to watch this resource instead of ClusterDeployment.
+
+In the future, we could migrate the current standalone openshift-install implementation to an OpenShiftInstallStrategy, which continues to use ClusterProvision for it's retry mechanism. To do this we would allow a parallel path to specify where you cannot use both installStrategyRef and provisioning sections together. Provisioning then becomes deprecated and after some period of time, could be removed.
+
+We could add an additional strategy which offers InstallConfig generation, and could be used by users who are ok with a simpler interface and don't want to specify an install config secret.
+
+HypershiftInstallStrategy would be another likely future implementation.
+
+### Examples
+
+```yaml
+apiVersion: hive.openshift.io/v1
+kind: AgentInstallStrategy
+metadata:
+  name: mycluster
+  namespace: hive
+spec:
+  stage: RetryProvisioning
+  clusterDeploymentRef:
+    name: mycluster
+  imageSetRef:
+    openshift-4.6
+  clusterMetadata:
```

Comment:
Why not have this in the Status? It's not user-declared state.

Reply:
We found in the past that if you want it persisted across a disaster recovery scenario (with velero or similar), it should go to spec if it cannot easily be reconstructed. This is why our "installed" flag is in spec as well.